lib/datatype/data-type-new.js

```diff
@@ -0,0 +1,206 @@
+    if (!validate(this.#schemaName, value)) {
+      // TODO: pass through errors from validate functions
+      throw new Error('Invalid value ' + value)
+    }
+    const nowDateString = generateDate()
+    /** @type {OmitUnion<MapeoDoc, 'versionId'>} */
+    const doc = {
+      ...value,
+      docId: generateId(),
+      createdAt: nowDateString,
+      updatedAt: nowDateString,
+      links: [],
+    }
+
+    // TS can't track the relationship between TDoc and TValue, so doc above is
+    // typed as MapeoDoc (without versionId) rather than as TDoc.
+    await this.#dataStore.write(/** @type {TDoc} */ (doc))
+    return this.getByDocId(doc.docId)
+  }
+
+  /**
+   * @param {string} docId
+   */
+  async getByDocId(docId) {
+    return deNullify(this.#sql.getByDocId.get({ docId }))
+  }
+
+  /** @param {string} versionId */
+  async getByVersionId(versionId) {
+    return this.#dataStore.read(versionId)
+  }
+
+  async getMany() {
+    return this.#sql.getMany.all().map((doc) => deNullify(doc))
+  }
+
+  /**
+   *
+   * @template {import('type-fest').Exact<TValue, T>} T
+   * @param {string | string[]} versionId
+   * @param {T} value
+   */
+  async update(versionId, value) {
+    const links = Array.isArray(versionId) ? versionId : [versionId]
+    const { docId, createdAt } = await this.#validateLinks(links)
+    /** @type {any} */
+    const doc = {
+      ...value,
+      docId,
+      createdAt,
+      updatedAt: new Date().toISOString(),
+      links,
+    }
+    await this.#dataStore.write(doc)
+    return this.getByDocId(docId)
+  }
+
+  /**
+   * Not yet implemented
+   * @param {string | string[]} versionId
+   */
+  async delete(versionId) {
+    const links = Array.isArray(versionId) ? versionId : [versionId]
+    const { docId, createdAt } = await this.#validateLinks(links)
+    /** @type {any} */
+    const doc = {
+      docId,
+      createdAt,
+      updatedAt: new Date().toISOString(),
+      links,
+      schemaName: this.#schemaName,
+      deleted: true,
+    }
+    await this.#dataStore.write(doc)
+    return this.getByDocId(docId)
+  }
+
+  /**
+   * Validate that existing docs with the given versionIds (links):
+   * - exist
+   * - have the same schemaName as this dataType
+   * - all share the same docId
+   * - all share the same createdAt
+   * Throws if any of these conditions fail, otherwise returns the validated
+   * docId and createAt datetime
+   * @param {string[]} links
+   * @returns {Promise<{ docId: MapeoDoc['docId'], createdAt: MapeoDoc['createdAt'] }>}
+   */
+  async #validateLinks(links) {
+    const prevDocs = await Promise.all(
+      links.map((versionId) => this.getByVersionId(versionId))
+    )
+    const { docId, createdAt } = prevDocs[0]
+    const areLinksValid = prevDocs.every(
+      (doc) => doc.docId === docId && doc.schemaName === this.#schemaName
+    )
+    if (!areLinksValid) {
+      throw new Error('Updated docs must have the same docId and schemaName')
+    }
```

Comment:
I was about to commit suggestion but hesitated. To me this makes it a bit harder to understand what the code is doing, because of the double negatives. The only reason for doing it is a very small performance improvement, but I think the performance here is insignificant - we're probably talking <0.001% of the time this method takes.